### ISSUE TITLE
self communication pattern

### ISSUE THREAD

**Issue by laroque** [driplineorg/dripline-python#105]
It is currently the case that an object in a dripline-python service cannot send requests to an endpoint in the same service. This is because the call stack send the request stops any python functions from processing the incoming request. There are a few pieces to resolving this:

1. We should re-evaluate the potential use of async, it may be possible to resolve blocking with those patterns (probably requires across the board refactoring to make everything async compatible)
2. The lower-level send methods should check if the target of the request is local, if so it should pass the request object to the request handling logic stack directly (rather than making a round trip to the broker). This places the resolution logic in the same call stack and avoid blocking.

There are some possible concerns:
1. The above could result in an infinitely recursive loop (which should be stopped by the python recursion depth limit). It is not clear superficially if this is something we need to detect and protect against, or just document and "let it fail" for a user to fix.
2. There could be cases where both a local and a remote consumer are expecting to receive a request, we need to make sure that the remote also gets it (for example, if I am running dl-mon to see all requests coming into some endpoint).
3. If the request is sent locally, the reply needs to also be processed locally such that the original caller receives the reply in the expected type and format. This also has the same concern as 2, namely that there could also be a non-local client expecting to receive the result.

**Comment by nsoblath** [comment]
I've tracked down the problem to a blocked C++ thread.  Here's a listing of the calls in two threads, the Main Thread (MT) and the Receiver Thread of the service.  The MT is waiting for AMQP messages, and the RT handles messages.  The issue is that the RT handles the first (CMD) request, and in the process of doing so sends a GET request that comes back to the service; before the GET request arrives, though, the RT blocks while waiting for the reply to the GET and therefore it cannot actually handle the GET.

![Screen Shot 2020-06-01 at 3 06 36 PM](https://user-images.githubusercontent.com/829991/83459970-2da34080-a41a-11ea-92e3-b7de215318f4.png)

One potential solution to this would be to have services use multiple receiver threads.  However, one concern with this would be the potential to have commands be handled out of order.  While this could happen in principle, it might not be a significant worry in practice depending on the rate at which messages are received and processed.

**Comment by nsoblath** [comment]
Regarding the point about local handling of requests, I agree that we'd need a way for both local and remote handling of the same message.  However, we also need to be careful that if a message is handled locally, it's not also handled as a remote message by the same endpoint.  

For example, let's say `service` is sending a message to RK `A`, and service has an endpoint bound to `A`.  `service` recognizes that it has `A` locally, so it directly calls the handler function.  It also sends the message to the broker in case there's any remote listeners.  The endpoint is one such listener, so it also gets sent the message from the broker and handles it a second time.  I'm not sure what the right solution is to this problem.

**Comment by laroque** [comment]
Yeah this case was one of the one I was intending to point out. I could imagine keeping a cache of requests received and responded to, which solves the duplicate response issue, but then we need some policy about how long to cache those to avoid infinite memory use growth. We'd also need to figure out what ID to cache and would have to solve the same for the reply.... Not insurmountable but needs to be done deliberately.

**Comment by raphaelcervantes** [comment]
I think getting this working would help me develop endpoints that run multiple SCPI commands. This is in reference to the slack conversation we had about the trouble I was having with a single endpoint running mutliple SCPI commands.

> If you want to do a sophisticated interaction with a device, (configure + trigger + retrieve data) you really only have two choices:
> 1) create a custom class to interact with it. In your case you could send the compound command as you’re already doing, but you’d need to parse out those extra +1;+1; from the front, before returning the data. … it may be possible to do this with a regex extraction per the format spime, but it is really abusing the way that class works and I wouldn’t be surprised to see it break in unforeseen ways in the future. In particular, this may be slow, dripline interactions are really expected to try and be quick
> 2) split the interaction across multiple endpoints and have whatever script/user is calling it first configure and then trigger, then poll to see if the data are ready, then get/log the data

**Comment by laroque** [comment]
@raphaelcervantes: realistically, I think there are two important points here:
1) The issue this ticket represents fixing is probably one of the largest feature gaps we're dealing with, and fixing it would be very useful to a number of projects. However, it really doesn't allow for any sort of quick fix, it will require that someone design a revision to the call logic and carefully treat the subtleties above.
2) Realistically I don't know that we have anyone with both the available time and expertise to work on this refactor.

I bring this up because I don't think it will make sense for you to wait and hope we'll be able to provide a solution on the timescale of your need. Also, assuming your priority is to get Orpheus running rather than contributing to dripline, you'll probably be better off following the pattern of SAG or locking and writing a non-reusable custom class to do what you want.

**Comment by raphaelcervantes** [comment]
Ok. Thank you for the heads up!
